{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting C:\\Users\\Cflhxb\\AppData\\Local\\anaconda3\\Lib\\CAMP.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile C:\\Users\\Cflhxb\\AppData\\Local\\anaconda3\\Lib\\CAMP.py\n",
    "    \n",
    "import datetime as dt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math as math \n",
    "import MathsUtilities as MUte\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import patches\n",
    "from scipy.optimize import curve_fit\n",
    "\n",
    "CampConstants = {\n",
    "    'k':-0.17,\n",
    "    'VIThreshold':1.0,\n",
    "    'TSThreshold':2.0,\n",
    "    'SlopeFLNvsTS':1.1,\n",
    "    'IntFLNvsTS':2.85,\n",
    "    'PpMax':16,\n",
    "    'PpMin':8,\n",
    "}\n",
    "\n",
    "colors = ['r','b','b','r']\n",
    "fills = ['w','b','w','r']\n",
    "\n",
    "def CalcdHS(Tt,Pp,HS,BasePhyllo,PhylloPpSens):\n",
    "    \"\"\"Calculate daily increase in Haun stage\n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        Pp: The current photoperiod\n",
    "        HS: Current Haun Stage\n",
    "        BasePhyllo: Base phyllochron at 16h Pp between 3 and 7 HS\n",
    "        PhylloPpSens: Photo period sensitivity of phyllochron.  Relative increase in phyllochron as 8h Pp\n",
    "    Returns:\n",
    "        Daily increas in Haun Stage\n",
    "    \"\"\"\n",
    "    HS = max(0,HS)\n",
    "    StageFactor = np.interp(HS,[0,2,2.0001,7,7.0001,12],[0.75,0.75,1.0,1.0,1.4,1.4],False)\n",
    "    PpFactor = 1 + PhylloPpSens * np.interp(Pp,[0,8,12,20],[1,1,0,0],False)\n",
    "    Phyllochron = BasePhyllo*StageFactor*PpFactor\n",
    "    return Tt/Phyllochron\n",
    "\n",
    "def CalcPpResponse(Pp, Base, Max):\n",
    "    \"\"\" Calculate value of a photoperiod sensitive variable between Max and Baee values\n",
    "    Args:\n",
    "        Pp: Photoperiod\n",
    "        Base: value below 8h Pp\n",
    "        Max: value above 16h Pp\n",
    "        dPB: delta base phyllochrons\n",
    "    Returns:\n",
    "        variable value dependent on photoperiod and adjusted for dHS\n",
    "    \"\"\"\n",
    "    if Pp <= CampConstants['PpMin']:\n",
    "        return Base\n",
    "    if (Pp > CampConstants['PpMin']) and (Pp < CampConstants['PpMax']):\n",
    "        return (Base + (Max-Base) * (Pp-CampConstants['PpMin'])/(CampConstants['PpMax']-CampConstants['PpMin']))\n",
    "    if (Pp >= CampConstants['PpMax']):\n",
    "        return Max\n",
    "\n",
    "def CalcBaseUpRegVrn1(Tt, dHS, BaseDVrn1):\n",
    "    \"\"\" Calculate upregulation of base Vrn1\n",
    "    \n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        dHS: Delta HaunStage\n",
    "        BaseDVrn1: coeffociennt for Base Vrn1 expression\n",
    "        \n",
    "    Returns:\n",
    "        delta BaseVrn1 representing the additional Vrn1 expression from base expression\n",
    "    \"\"\"\n",
    "    if Tt < 0: \n",
    "        BaseDVrn1 = 0\n",
    "    return BaseDVrn1 * dHS\n",
    "\n",
    "def CalcColdUpRegVrn1(Tt,dHS, MaxApDev1, k):\n",
    "    \"\"\" Upregulation of Vrn1 from cold.  Is additional to base vrn1\n",
    "        BaseDVrn1 in seperate calculation otherwise te same as Brown etal 2013\n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        dHS: Delta Base Phyllochrons\n",
    "        MaxApDev1: coefficient for Maximum upregulation of Vrn1\n",
    "        k: The exponential coefficient determining temperature response\n",
    "        \n",
    "    Returns:\n",
    "        delta ColdVrn1 representing the additional Vrn1 expression from cold upregulation\n",
    "    \"\"\"\n",
    "    UdVrn1 = MaxApDev1 * np.exp(k*Tt)\n",
    "    if (Tt < 20):\n",
    "        return UdVrn1 * dHS\n",
    "    else:\n",
    "        return -5\n",
    "        \n",
    "def calcTSHS(FLN,IntFLNvsTSHS,SlopeFLNvsTS):\n",
    "    \"\"\"Haun stage timing of terminal spikelet \n",
    "       Inverts equation 5 from Brown etal 2013 FLN =  2.85 + 1.1*TSHS and converts it to base phyllochrons.\n",
    "       Note the intercept differs, was typeo on publication\n",
    "       Intercept has been made variable for as needs to be lower for some very fast varieties\n",
    "    Args:\n",
    "        FLN: The final leaf number\n",
    "        IntFLNvsTSHS: The intercept of the regression between FLN and TSHS\n",
    "    Returns:\n",
    "        Estimation of number of base phyllochrons to terminal spikelet\n",
    "    \"\"\"\n",
    "    return (FLN - IntFLNvsTSHS)/SlopeFLNvsTS\n",
    "\n",
    "def CAMPmodel(Out, Day, Tt, Pp, Params, Consts, TtEmerge):\n",
    "    \"\"\" The Cereal Anthesis Molecular Phenology model.\n",
    "        Based on the ideas presented in Brown etal 2014 (Annals of Botany)\n",
    "        Alterations made replacing Vrn4 notion with methalation of Vrn1, allowing Vrn3 to act before VI and changint the pattern for Vrn2 expression\n",
    "        Other alterations to implement working code base\n",
    "    Args:\n",
    "        Out: \"FLN\" or None.  If \"FLN\" will only return estimated FLN else will return full dataframe with daily state variable values\n",
    "        Day: List, 1:EndDay representing the timesteps in model run\n",
    "        Tt: List of same length as Day, representing daily temperature\n",
    "        Pp: List of same length as Day, representing daily photoperiod\n",
    "        Params: Dict with genotype parameters fitted by CalcCultivarVrnCoeffs() \n",
    "        Consts: Dict with crop specific constants,\n",
    "        TtEmerge: Thermaltime from sowing to emergence\n",
    "    \"\"\"\n",
    "    # Set up Data structure and initialise values\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    DF = pd.DataFrame(index = [0], columns = ['Day','Tt','Pp'])\n",
    "    IsGerminated = False\n",
    "    IsEmerged = False\n",
    "    IsMethalating = False\n",
    "    IsVernalised = False\n",
    "    IsReproductive = False\n",
    "    IsAtFlagLeaf = False\n",
    "    DF.loc[0,'Day'] = 0\n",
    "    DF.loc[0,'Tt'] = Tt[0]\n",
    "    DF.loc[0,'RelCold'] = 0\n",
    "    DF.loc[0,'Pp'] = Pp[0]\n",
    "    DF.loc[0,'fPP'] = 0\n",
    "    DF.loc[0,'Stage'] = 'Germination'\n",
    "    DF.loc[0,'IsGerminated'] =True\n",
    "    DF.loc[0,'IsEmerged'] = False\n",
    "    DF.loc[0,'IsVernalised'] = False\n",
    "    DF.loc[0,'IsReproductive'] = False\n",
    "    DF.loc[0,'IsAtFlagLeaf'] = False\n",
    "    DF.loc[0,'HS'] = -TtEmerge/Params.BasePhyllo\n",
    "    DF.loc[0,'dHS'] = 0\n",
    "    DF.loc[0,'AccumTt'] = Tt[0]\n",
    "    DF.loc[0,'VrnB'] = 0\n",
    "    DF.loc[0,'dVrnB'] = 0\n",
    "    DF.loc[0,'MaxApDev'] = 0\n",
    "    DF.loc[0,'dMaxApDev'] = 0\n",
    "    DF.loc[0,'Cold'] = 0\n",
    "    DF.loc[0,'dCold'] = 0\n",
    "    DF.loc[0,'ApDev'] = 0\n",
    "    DF.loc[0,'MaxVrn2'] = 0\n",
    "    DF.loc[0,'Vrn2'] = 0.0\n",
    "    DF.loc[0,'AccumSPp'] = 0.0\n",
    "    #DF.loc[0,'Vrn2ef'] = 0.0\n",
    "    DF.loc[0,'Vrn1'] = 0.0\n",
    "    DF.loc[0,'dVrn1'] = 0\n",
    "    DF.loc[0,'Vrn3'] = 0.0\n",
    "    DF.loc[0,'dVrn3'] = 0\n",
    "    DF.loc[0,'VIHS'] = 0\n",
    "    DF.loc[0,'TSHS'] = 0\n",
    "    DF.loc[0,'FLN'] = Consts['IntFLNvsTS']\n",
    "    \n",
    "    d = 1\n",
    "    \n",
    "    # Model daily loop\n",
    "    # ^^^^^^^^^^^^^^^^\n",
    "    while (IsAtFlagLeaf == False) and (d < Day[-1]):\n",
    "        ColdYesterday = 0\n",
    "        ColdYesterday = DF.loc[d-1,'Cold']\n",
    "        DF.loc[d,:] = DF.loc[d-1,:] #Carry yesterdays values over to today\n",
    "        DF.loc[d,'Stage'] = np.nan\n",
    "            \n",
    "        DF.loc[d,'Day'] = d\n",
    "        # Set daily environment variables\n",
    "        DF.loc[d,'Tt'] = Tt[d]\n",
    "        DF.loc[d,'Pp'] = Pp[d]\n",
    "        #Zero set deltas\n",
    "        DF.loc[d,'dHS'] = 0\n",
    "        DF.loc[d,'dVrnB'] = 0\n",
    "        DF.loc[d,'dMaxApDev'] = 0\n",
    "        DF.loc[d,'dCold'] = 0\n",
    "        DF.loc[d,'Vrn2'] = 0\n",
    "        DF.loc[d,'dVrn1'] = 0\n",
    "        DF.loc[d,'dVrn3'] = 0\n",
    "        \n",
    "                \n",
    "        DF.loc[d,'AccumTt'] = DF.loc[d,'AccumTt'] + DF.loc[d,'Tt']\n",
    "        \n",
    "        PropnOfDay = 1.0\n",
    "        if (DF.loc[d,'AccumTt'] > TtEmerge) and (IsEmerged==False):\n",
    "            IsEmerged = True\n",
    "            DF.loc[d,'Stage'] = 'Emergence'\n",
    "            PropnOfDay = (DF.loc[d,'AccumTt'] - TtEmerge)/DF.loc[d,'Tt'] # Calculate fraction of emergence days Tt that is not used for emergence\n",
    "        \n",
    "        # Calculate daily Haun Stage changes\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if IsEmerged==False: # Crop not yet emerged \n",
    "            EmergDurationFactor = 1\n",
    "            if (DF.loc[d,'AccumTt'] > 90):  # Calculate EmergenceDurationFactor to slow accumulation of HS if emergence is taking a long time.  This slows Vrn1 expression under slow emergence and strange responses to delayed sowing\n",
    "                EmergDurationFactor = np.exp(-0.015 * (DF.loc[d,'AccumTt']-90))\n",
    "            DF.loc[d,'dHS'] = DF.loc[d,'Tt']/Params.BasePhyllo  * EmergDurationFactor\n",
    "            DF.loc[d,'fPP'] = 0\n",
    "            DF.loc[d,'Vrn2'] = 0\n",
    "            \n",
    "        else: # Crop emerged\n",
    "            # Calculate delta haun stage\n",
    "            DF.loc[d,'dHS'] = CalcdHS(DF.loc[d,'Tt'],DF.loc[d,'Pp'],DF.loc[d,'HS'],Params.BasePhyllo,Params.PhylloPpSens)\n",
    "            \n",
    "        #increment HS\n",
    "        DF.loc[d,'HS'] += DF.loc[d,'dHS']\n",
    "        \n",
    "        # Calculate Vrn gene expression\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if (IsReproductive==False):\n",
    "            if (IsVernalised==False): #set as rates and factors for vegetative phase\n",
    "                DF.loc[d,'dVrnB'] = Params.rVrnBVeg * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'dMaxApDev'] = Params.rApDevMVeg * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'PpVrn3Fact'] = Params.rVrn3Veg\n",
    "            else:\n",
    "                 #set as rates and factors for early reproductive phase\n",
    "                DF.loc[d,'dVrnB'] = Params.rVrnBER * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'dMaxApDev'] = Params.rApDevMER * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'PpVrn3Fact'] = Params.rVrn3ER\n",
    "            #Increment BaseVern and MaxVern\n",
    "            DF.loc[d,'VrnB'] += DF.loc[d,'dVrnB']\n",
    "            DF.loc[d,'MaxApDev'] = min(2,DF.loc[d,'MaxApDev'] + DF.loc[d,'dMaxApDev'])\n",
    "\n",
    "            # Calculate daily cold response\n",
    "            if (IsVernalised==False):\n",
    "                DF.loc[d,'RelCold'] = CalcColdUpRegVrn1(DF.loc[d,'Tt'],1,1,Consts['k'])\n",
    "                DF.loc[d,'dCold'] = DF.loc[d,'RelCold'] * DF.loc[d,'dVrnB'] * Params.rVrn1\n",
    "                DF.loc[d,'Cold'] = max(0.0,DF.loc[d,'Cold'] + DF.loc[d,'dCold'])\n",
    "                if DF.loc[d,'Cold'] > Params.MethalationThreshold:\n",
    "                    DF.loc[d,'dVrn1'] = max(0,(DF.loc[d,'Cold'] - Params.MethalationThreshold)-(ColdYesterday - Params.MethalationThreshold))\n",
    "            # Increment Vrn 1\n",
    "            DF.loc[d,'Vrn1'] += DF.loc[d,'dVrn1']\n",
    "\n",
    "            # Calcualte expression of photoperiod sensitive genes\n",
    "            if (IsEmerged==True):  # Photoperiod sensitive genes only express after emergence\n",
    "                DF.loc[d,'fPP'] = CalcPpResponse(DF.loc[d,'Pp'],0,1) #relative Pp, scaled between 0 at lower threshold and 1 at upper threshold\n",
    "                DF.loc[d,'AccumSPp'] += DF.loc[d,'dVrnB'] * (1-DF.loc[d,'fPP'])  #accumulate short day exposure for reducing rVrn2\n",
    "                DF.loc[d,'MaxVrn2'] = max(0,(Params.rVrn2 * DF.loc[d,'fPP']) - DF.loc[d,'AccumSPp'])\n",
    "                DF.loc[d,'Vrn2'] = max(0,DF.loc[d,'MaxVrn2'] - (DF.loc[d,'VrnB'] + DF.loc[d,'Vrn1'])) \n",
    "                if (DF.loc[d,'Vrn2'] == 0): # express vrn3 relative to Pp if effective Vrn2 is down regulated to zero\n",
    "                    DF.loc[d,'dVrn3'] = (DF.loc[d,'PpVrn3Fact']-1) * DF.loc[d,'fPP'] * DF.loc[d,'dVrnB']\n",
    "                #Increment Vrn3\n",
    "            DF.loc[d,'Vrn3'] += DF.loc[d,'dVrn3']\n",
    "\n",
    "            # Increment todays Vrn expression values using deltas just calculated\n",
    "            if (IsVernalised==False): #If not vernalised need to include Vrn2 and constrain to maxrate\n",
    "                DF.loc[d,'ApDev'] = min(DF.loc[d,'MaxApDev'],max(0,DF.loc[d,'VrnB'] + DF.loc[d,'Vrn1'] + DF.loc[d,'Vrn3'] - DF.loc[d,'MaxVrn2'])) \n",
    "            else:\n",
    "                DF.loc[d,'ApDev'] = DF.loc[d,'ApDev'] + DF.loc[d,'dVrnB'] + DF.loc[d,'dVrn3']\n",
    "\n",
    "        # Set Haun stage variables\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if IsVernalised == False:\n",
    "            DF.loc[d,'VIHS'] = DF.loc[d,'HS']\n",
    "        if IsReproductive == False:\n",
    "            DF.loc[d,'TSHS'] = DF.loc[d,'HS']\n",
    "            DF.loc[d,'FLN'] = Consts['IntFLNvsTS'] + Consts['SlopeFLNvsTS'] * DF.loc[d,'TSHS']\n",
    "\n",
    "        # Finally determine phenological stage\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        # Vernalisation saturation occurs when Vrn1 > the vernalisation threshold and Vrn2 expression is zero\n",
    "        if (IsEmerged==True) and (DF.loc[d,'ApDev'] >= Consts['VIThreshold']) and (IsVernalised==False):\n",
    "            IsVernalised  = True\n",
    "            DF.loc[d,'Stage'] = 'Vern Sat'\n",
    "            #DF.loc[d,'Vrn1'] = 0\n",
    "        \n",
    "        if (DF.loc[d,'ApDev'] >=  Consts['TSThreshold']) and (IsReproductive == False):\n",
    "            IsReproductive = True;\n",
    "            DF.loc[d,'Stage'] = 'Term Spike'\n",
    "        \n",
    "        #Work out if Flag leaf has appeared.\n",
    "        if DF.loc[d,'HS'] >= DF.loc[d,'FLN']:\n",
    "            IsAtFlagLeaf = True    \n",
    "            DF.loc[d,'Stage'] = 'FlagLeaf'\n",
    "    \n",
    "        #Add states to dataframe\n",
    "        IsDayOfEmergence = False\n",
    "        \n",
    "        DF.loc[d,'IsGerminated'] = IsGerminated\n",
    "        DF.loc[d,'IsEmerged'] = IsEmerged\n",
    "        DF.loc[d,'IsVernalised'] = IsVernalised\n",
    "        DF.loc[d,'IsReproductive'] = IsReproductive\n",
    "        DF.loc[d,'IsAtFlagLeaf'] = IsAtFlagLeaf\n",
    "    \n",
    "        # Increment day\n",
    "        d += 1\n",
    "        \n",
    "    if (Out == 'FLN') and (IsAtFlagLeaf == True):\n",
    "        return DF.iloc[-1,:]['FLN']\n",
    "    else:\n",
    "        return DF\n",
    "    \n",
    "def deriveVrnParams(c):\n",
    "    data = pd.Series(dtype=np.float64)\n",
    "    # Calculate the accumulated base phyllochrons at terminal spikelet for each treatment\n",
    "    for pv in ['WS','CL','CS','WL']:\n",
    "        data['TS_'+pv] = calcTSHS(c['FLN_'+pv],CampConstants['IntFLNvsTS'],CampConstants['SlopeFLNvsTS'])\n",
    "    # Base Phyllochron\n",
    "    data['BasePhyllo'] = c['BasePhyllo']\n",
    "    # Phyllochron photoperiod sensitivity.  #This is not used in calculations but added so can be checked in outputs\n",
    "    data['PhylloPpSens'] = c['PhylloPpSens']\n",
    "    # Base Phyllochron duration of the Emergence Phase\n",
    "    data['EmergDurat'] = c['TtEmerge']/data['BasePhyllo']\n",
    "    # Base Phyllochron duration of vernalisation treatment\n",
    "    data['VernTreatDurat'] = (c['VrnTreatDuration']*c['VrnTreatTemp'])/data['BasePhyllo']\n",
    "    # Accumulated Base Phyllochrons when vernalisation treatment ended \n",
    "    data['EndVernTreat'] =  -data['EmergDurat'] + data['VernTreatDurat']\n",
    "    # The minimum accumulated haunstage at which Vern saturation can occur\n",
    "    data['MinVI'] = 1.1\n",
    "    # The minimum haun stage duration from vern saturation to terminal spikelet\n",
    "    data['MinVI->TS'] = min(3,data['TS_CL']-data['MinVI'])  #From Lincoln controlled environment data\n",
    "    \n",
    "    # Calculate the accumulated base phyllochrons at vernalisation saturation for each treatment\n",
    "    data['VI_CL'] = data['TS_CL'] - data['MinVI->TS']   \n",
    "    data['VI_WL'] = max(data['TS_WL'] - data['MinVI->TS'],data['VI_CL']) #Constrained so not earlier than the CL treatment\n",
    "    data['VI_CS'] = min(data['VI_CL'],data['TS_CS']-data['MinVI->TS']) # Assume happens at the same time as VI_CL but not sooner than TS and MinVI->TS would allow \n",
    "    data['VI->TS_CS'] = data['TS_CS'] - data['VI_CS']\n",
    "    data['VI_WS'] = max(data['TS_WS'] - data['VI->TS_CS'],data['VI_CL']) #Constrained so not earlier than the CL treatment\n",
    "    \n",
    "    # Calculate base and maximum rates and Pp sensitivities\n",
    "    # Base Vrn delta during vegetative phase.  Assuming base Vrn expression starts at sowing and reaches 1 at VI for the WS treatment where no cold or Vrn3 (short Pp) to upregulate \n",
    "    data['rVrnBVeg'] = 1/(data['VI_WS']+data['EmergDurat'])\n",
    "    # The fastest rate that Vrn can accumulate to reach saturation (VI).\n",
    "    data['rApDevMVeg'] = 1/(data['VI_CS']+data['EmergDurat'])\n",
    "    # Base Vrn delta during early reproductive phase.  Assuming Vrn expression increases by 1 between VI and TS and proceeds at a base rate where not Vrn3 upregulation (short PP)\n",
    "    data['rVrnBER'] = 1/(data['TS_WS']-data['VI_WS'])\n",
    "    # The maximum Vrn delta during the early reproductive phase.  Assuming Vrn increases by 1 bewtween VI and TS and proceeds at the maximum rate under long photoperiod\n",
    "    data['rApDevMER'] = 1/data['MinVI->TS']\n",
    "    # The relative increase in delta Vrn cuased by Vrn3 expression under long photoperiod during early reproductive phase\n",
    "    data['rVrn3ER'] = data['rApDevMER']/data['rVrnBER']\n",
    "    # The relative increase in delta Vrn1 caused by Vrn3 expression under long Pp during vegetative phase.  Same as rVrn3ER unless VI_WL is small\n",
    "    data['rVrn3Veg'] = (1/data['VI_WL'])/data['rVrnBVeg'] \n",
    "    data['rVrn3Veg'] = max(data['rVrn3ER'],data['rVrn3Veg']) \n",
    "    \n",
    "    # Calculate Maximum Vrn2 expression under long photoperiods\n",
    "    # The rate of Vrn expression before VI under long Pp when Vrn2 = 0 is the product of Vrn3 and base Vrn\n",
    "    # Therefore the haun stage duration when effective baseVrn x Vrn3 is up regulating during the vegetative phase under long photoperiod without vernalisation is a funciton of the VI target (1) and the rate\n",
    "    data['URVrn3HSWL'] = 1/(data['rVrnBVeg'] * data['rVrn3Veg'])\n",
    "    # The accumulated Haun stages when Vrn2 expression is ended and effective baseVrn x Vrn3 expression starts under long Pp without vernalisation is vrn1xVrn3Durat prior to VI\n",
    "    data['DRVrn2HSWL'] = max(0,data['VI_WL'] - data['URVrn3HSWL']) \n",
    "    # The Vrn2 Expression that must be matched by Vrn1 before effective Vrn3 expression starts.  \n",
    "    data['rVrn2'] = (data['DRVrn2HSWL']+data['EmergDurat']) * data['rVrnBVeg']\n",
    "    \n",
    "    # Calculate cold upregulation of Vrn1\n",
    "    # The amount of protaganistic vrn expression required to reach VI\n",
    "    data['VrnPVI_CL'] = CampConstants['VIThreshold'] + data['rVrn2']\n",
    "    # The amount of Vrn expression at base rate in the CL treatment between sowing and when Vrn2 is suppressed \n",
    "    data['VrnBVeg_CL'] = (data['EmergDurat']+data['VI_CL']) * data['rVrnBVeg']\n",
    "    # The amount of persistant (methalated) Vrn1 upregulated due to cold at the end of the vernalisation treatment for CL\n",
    "    # Is VrnPVI_CL less base vrn expression up to VI.  Assuming Vrn3 expression prior to VI is neglegable in this case \n",
    "    data['Vrn1_CL'] = data['VrnPVI_CL'] - data['VrnBVeg_CL'] \n",
    "    # Cold threshold required before Vrn1 starts methalating.  Based on Brooking and Jamieson data the lag duration is the same length as the duration of response up to full vernalisation\n",
    "    # Therefore we assume the methalation threshold must be the same size as the amount of vrn1 that is required to give full persistend vernalisation response.\n",
    "    data['MethalationThreshold'] = data['Vrn1_CL']\n",
    "    # Haun stage duration of cool treatment, may be less than treatment duration if treatment goes past VI\n",
    "    data['CTHS'] =  min(data['VernTreatDurat'],data['VI_CL']+data['EmergDurat'])\n",
    "    # The rate of Vrn1 expression under cold treatment calculated from the amopunt of cold vrn1 up regulation apparante plus the methalation threshold\n",
    "    data['DVrn1_CL'] = (data['Vrn1_CL']+data['MethalationThreshold'])/data['CTHS'] \n",
    "    # The upregulation of Vrn1 expression above base rate at 0oC\n",
    "    data['DVrn1Max'] = data['DVrn1_CL']/ np.exp(c['k'] * c['VrnTreatTemp']) \n",
    "    # The relative increase in delta Vrn1 caused by cold upregulation of Vrn1\n",
    "    data['rVrn1'] = data['DVrn1Max'] /data['rVrnBVeg']\n",
    "                                 \n",
    "    return data\n",
    "\n",
    "def plotFLNs(CL,CS,WS,WL,Axis,fs,maxFLN=20):\n",
    "    width = 0.4\n",
    "    ind = np.arange(4) + width\n",
    "    plt.bar(ind+.4, [CL,CS,WS,WL], width,\n",
    "            edgecolor=['b','b','r','r'], color = ['b','w','w','r'],\n",
    "            linewidth=3,alpha=0.3)\n",
    "    plt.ylabel('Final Leaf Number',fontsize=12)\n",
    "    plt.tick_params(labelsize=fs)\n",
    "    Axis.set_xticks(ind+width)\n",
    "    Axis.set_xticklabels(['$CL$','$CS$', '$WS$', '$WL$'])\n",
    "    plt.ylim(0,maxFLN)\n",
    "    plt.xlim(0.5,4.1)\n",
    "    Axis.spines['right'].set_visible(False)\n",
    "    Axis.spines['top'].set_visible(False)\n",
    "    plt.tick_params(axis='x', which='both', bottom=True,top=False, labelbottom=True)\n",
    "    plt.tick_params(axis='y', which='both', left=True,right=False, labelleft=True)\n",
    "\n",
    "def addArrow(Xs,Ys,col='k'):\n",
    "    bx = Xs[0]\n",
    "    by = Ys[0]\n",
    "    dx = Xs[1] \n",
    "    dy = Ys[1]\n",
    "    prop = dict(arrowstyle=\"-|>,head_width=0.2,head_length=0.5\",color=col,lw=1,ls=':',\n",
    "            shrinkA=0,shrinkB=0)\n",
    "    plt.annotate(\"\",xy=(dx,dy),xytext=(bx,by), arrowprops=prop)\n",
    "    \n",
    "def plotLNlines(CL,CS,WS,WL,fs):\n",
    "    MinLN = CL\n",
    "    PpLN = max(0,CS - CL)\n",
    "    CvLN = max(0,WS - MinLN - PpLN)\n",
    "    PvLN = WL - MinLN - CvLN\n",
    "    \n",
    "    #MinLN\n",
    "    addArrow([.6,3],[MinLN]*2)\n",
    "    plt.plot([0.6]*2,[0,MinLN],'-',color='k',lw=4)\n",
    "    plt.text(0.7,CL-2,'$MinLN=$'+\"{:.1f}\".format(MinLN),fontsize=fs,verticalalignment='center')\n",
    "    #PpLN\n",
    "    if PpLN>0.5:\n",
    "        addArrow([1.6,2.6],[MinLN+PpLN]*2,'darkorange')\n",
    "        plt.plot([1.6,1.6],[CS,CL],'-',color='darkorange',lw=4)\n",
    "        plt.text(1.5,max((CL+CS)/2,CL+.75),'$PpLN=$'+\"{:.1f}\".format(PpLN),\n",
    "                 fontsize=fs,verticalalignment='center',horizontalalignment='right',color='darkorange')\n",
    "    else:\n",
    "        plt.text(1.3,(CL+CS)/2,'$PpLN=0$',fontsize=fs,verticalalignment='bottom',color='darkorange')\n",
    "    #CvLN\n",
    "    if CvLN > 0.5:\n",
    "        plt.plot([2.6,2.6],[CS,MinLN+PpLN+CvLN],'-',color='darkblue',lw=4)\n",
    "        plt.text(2.5,max((CS+WS)/2,CS+.75),'$CvLN=$'+\"{:.1f}\".format(CvLN),\n",
    "                 fontsize=fs, verticalalignment='center',horizontalalignment='right',color='darkblue')\n",
    "        \n",
    "        plt.plot([3,3],[MinLN,MinLN+CvLN],'-',color='darkblue',lw=4)\n",
    "        addArrow([2.6,3],[MinLN+PpLN+CvLN*.5,MinLN+CvLN*.5],'darkblue')\n",
    "        addArrow([3,3.2],[MinLN + CvLN]*2,'darkblue')\n",
    "    else:\n",
    "        plt.text(2.1,(CS+WS)/2,'$CvLN=0$',fontsize=fs, verticalalignment='bottom',color='darkblue')\n",
    "    #PvLN\n",
    "    if ((PvLN > 0.5)or(PvLN<-0.5)):\n",
    "        plt.plot([3.2,3.2],[CL+CvLN,WL],'-',color='forestgreen',lw=4)\n",
    "        plt.text(3.3,max(CL+CvLN + PvLN/2,WL+0.75),'$PvLN=$'+\"{:.1f}\".format(PvLN),\n",
    "                 fontsize=fs, verticalalignment='center',color = 'darkgreen')\n",
    "    else:\n",
    "        plt.text(3.3,WL+0.75,'$PvLN=0$',fontsize=fs, verticalalignment='center',color = 'darkgreen')\n",
    "    addArrow([3.6,3.2],[WL,WL],'darkgreen')\n",
    "    \n",
    "def ExpressionPlot(data,c):\n",
    "    lowestTS = 16\n",
    "    highestTS = 0\n",
    "    pos = 0\n",
    "    #Plot TS for all treats at apparent Vrn1 expression of 2.0\n",
    "    for pv in ['WS','CL','CS','WL']:\n",
    "        #plt.plot(CampInputs.loc[c,'FLN_'+pv],2.5,'*',mfc = fills[pos],mec=colors[pos],ms=10)\n",
    "        plt.plot([data['TS_'+pv]],[2],'o',mfc = fills[pos],mec=colors[pos],ms=10)\n",
    "        lowestTS = min(lowestTS,data['TS_'+pv])\n",
    "        highestTS = max(highestTS,data['TS_'+pv])\n",
    "        plt.plot([data['TS_'+pv]],[2.1],'v',color = 'k',ms=7)\n",
    "        plt.plot([data['TS_'+pv]]*2,[2.1,2.2],'-',color = 'k',ms=7)\n",
    "        pos+=1\n",
    "    plt.plot([lowestTS,highestTS * 1.2],[2.2,2.2],'-',color = 'k',ms=7)\n",
    "    plt.text(highestTS * 1.25, 2.2,'TS',horizontalalignment='center')\n",
    "    \n",
    "    #Plot VI_CL treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_CL']],[1],'o',mec='b',mfc='b',ms=10)\n",
    "    plt.plot([data['VI_CL'],data['TS_CL']],[2,2],'--',color='k')\n",
    "    plt.text((data['VI_CL']+data['TS_CL'])/2,2.05,'MinVI->TS',horizontalalignment='center')\n",
    "    plt.plot([data['VI_CL']]*2,[1,2],'--',color='k')\n",
    "    plt.text(data['VI_CL'],1.05,'VI_CL',horizontalalignment='center',verticalalignment='bottom')\n",
    "    #Plot VI_LN treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_WL'],data['TS_WL']],[2,2],'--',color='k')\n",
    "    plt.text((data['VI_WL']+data['TS_WL'])/2,2.05,'MinVI->TS',horizontalalignment='center')\n",
    "    plt.plot([data['VI_WL']]*2,[1,2],'--',color='k')\n",
    "    plt.plot([data['VI_WL']],[1],'o',mec='r',mfc='r',ms=10)\n",
    "    plt.text(data['VI_WL'],1.05,'VI_WL',horizontalalignment='center',verticalalignment='bottom')\n",
    "    #Plot VI_CS treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_CS']],[1],'o',mec='b',mfc='w',ms=10)\n",
    "    plt.text(data['VI_CS'],0.95,'VI_CS',horizontalalignment='center',verticalalignment='top')\n",
    "    plt.plot([data['VI_CS'],data['TS_CS']],[1.8,1.8],':',color='k')\n",
    "    plt.plot([data['VI_CS'],data['VI_CS'],np.nan,data['TS_CS'],data['TS_CS']],[1,1.8,np.nan,1.8,2],':',color='k')\n",
    "    plt.text((data['VI_CS']+data['TS_CS'])/2,1.85,'VI->Ts_S',horizontalalignment='center')\n",
    "    #Plot VI_WS treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_WS']],[1],'o',mec='r',mfc='w',ms=10)\n",
    "    plt.text(data['VI_WS'],0.95,'VI_WS',horizontalalignment='center',verticalalignment='top')\n",
    "    plt.plot([data['VI_WS'],data['TS_WS']],[1.65,1.65],':',color='k')\n",
    "    plt.plot([data['VI_WS'],data['VI_WS'],np.nan,data['TS_WS'],data['TS_WS']],[1,1.65,np.nan,1.65,2],':',color='k')\n",
    "    plt.text((data['VI_WS']+data['TS_WS'])/2,1.7,'VI->Ts_S',horizontalalignment='center')\n",
    "    \n",
    "    #Plot base Vern1 Rate Vegetative\n",
    "    plt.plot([-data['EmergDurat'],data['VI_WS']],[0,(data['VI_WS']+data['EmergDurat'])*data['rVrnBVeg']],'-',color='g')\n",
    "    MidP = (-data['EmergDurat']+data['VI_WS'])/2\n",
    "    plt.plot([MidP,MidP+1],[0.5,0.5],'--',color='g')\n",
    "    plt.text(MidP+.5,0.53,'rVrnBVeg',color='g')\n",
    "    \n",
    "    #Plot max Vrn1 rate vegetative\n",
    "    plt.plot([-data['EmergDurat'],data['VI_CS']],[0,(data['VI_CS']+data['EmergDurat'])*data['rApDevMVeg']],'-',color='gold')\n",
    "    MidP = (data['VI_CL']+data['TS_CL'])/2\n",
    "    plt.plot([MidP,MidP+1],[1.5,1.5],'-',color='gold')\n",
    "    plt.text(MidP+.5,.53,'maxDVrnVeg',color='gold')\n",
    "    \n",
    "    #Plot base Vern1 Rate Early reproductive\n",
    "    plt.plot([data['VI_WS'],data['TS_WS']],[1,1+(data['TS_WS'] - data['VI_WS'])*data['rVrnBER']],'-',color='g')\n",
    "    MidP = (data['VI_WS']+data['TS_WS'])/2\n",
    "    plt.plot([MidP,MidP+1],[1.5,1.5],'--',color='g')\n",
    "    plt.text(MidP+.5,1.53,'rVrnBER',color='g')\n",
    "    \n",
    "    #Plot max Vrn1 ER\n",
    "    plt.plot([data['VI_CL'],data['TS_CL']],[1,1+(data['TS_CL'] - data['VI_CL'])*data['rApDevMER']],'-',color='gold')\n",
    "    MidP = (data['VI_CL']+data['TS_CL'])/2\n",
    "    plt.plot([MidP,MidP+1],[1.5,1.5],'-',color='gold')\n",
    "    plt.text(MidP+.5,1.53,'rApDevMER',color='gold')\n",
    "    \n",
    "    # Plot end of vernalisation timing\n",
    "    data['endVern'] = min(data['EndVernTreat'],data['VI_CL'])\n",
    "    plt.plot([data['endVern']]*2,[-1,2],'--',color='brown')\n",
    "    plt.text(data['endVern'],-0.1,'endVern',horizontalalignment='center',color='brown')\n",
    "    \n",
    "    #Extrapolate vrn1 back from VI_WL to show when it started effective expression\n",
    "    plt.plot([data['DRVrn2HSWL'],data['VI_WL']],[0,data['URVrn3HSWL'] * (data['rVrnBVeg'] * data['rVrn3Veg'])],'--',color='grey')\n",
    "    MidP = data['DRVrn2HSWL']+data['URVrn3HSWL']*0.1\n",
    "    plt.plot([MidP,MidP+1],[0.1]*2,'--',color='grey')\n",
    "    plt.text(MidP+.2,0.15,'dvrnxVrn3',color='grey')\n",
    "    \n",
    "    #Extrapolate BaseVrn1 back to MinVI to show how much Vrn1 was required to overcome Vrn2\n",
    "    plt.plot([0,data['DRVrn2HSWL']],[data['rVrn2'],0],'--',color='Orange')\n",
    "    plt.text(0,min(2.15,data['rVrn2']),'Vrn2',color='orange')\n",
    "    \n",
    "    #The amount of Vrn1 required to reach VI is light is 1 + rVrn2\n",
    "    plt.plot([0,data['VI_CL']],[1+data['rVrn2']]*2,'--',color='k')\n",
    "    plt.text(data['VI_CL']-.2,1+data['rVrn2']+.03,'Vrn Target',horizontalalignment='right',verticalalignment='bottom')\n",
    "    #the amount required from cold up regulation is the above less base Vrn1 until the end of Vrn2 less base * Vrn3 between end vrn2 and VI\n",
    "    plt.plot([0,data['VI_CL']],[1+data['rVrn2']-data['VrnBVeg_CL']]*2,'-',color='g')\n",
    "    plt.text(data['VI_CL']-.2,data['Vrn1_CL']+.03,'less baseVrn',horizontalalignment='right',verticalalignment='bottom')\n",
    "    \n",
    "    #RelMethTime = camp.CampConstants['MethalationThreshold']/(data['Vrn1_CL']+camp.CampConstants['MethalationThreshold'])\n",
    "    StartMethHS = -data['EmergDurat'] + data['CTHS'] * 0.5 \n",
    "    plt.plot([StartMethHS,data['endVern']],[0,data['Vrn1_CL']],'-',color='cyan')\n",
    "    plt.text(-data['EmergDurat'],data['Vrn1_CL'],'Vrn1_CL',horizontalalignment='right',rotation=90,verticalalignment='top',color='c')\n",
    "    \n",
    "    plt.ylabel('Apparent Vrn Expression')\n",
    "    plt.xlabel('Haun Stage')\n",
    "    plt.ylim(-.2,2.5)\n",
    "    plt.xlim(-2,20)\n",
    "    plt.text(19,0.05,c,horizontalalignment='right')\n",
    "    \n",
    "def plotVITS(data, level, c):\n",
    "    lowestTS = 16\n",
    "    highestTS = 0\n",
    "    pos = 0\n",
    "    #Plot TS for all treats at apparent Vrn1 expression of 2.0\n",
    "    for pv in ['WS','CL','CS','WL']:\n",
    "        #plt.plot(CampInputs.loc[c,'FLN_'+pv],2.5,'*',mfc = fills[pos],mec=colors[pos],ms=10)\n",
    "        plt.plot([data['TS_'+pv]],[2],'o',mfc = fills[pos],mec=colors[pos],ms=10)\n",
    "        plt.text(data['TS_'+pv],2.25,pv,horizontalalignment='center')\n",
    "        lowestTS = min(lowestTS,data['TS_'+pv])\n",
    "        highestTS = max(highestTS,data['TS_'+pv])\n",
    "        plt.plot([data['TS_'+pv]],[2.1],'v',color = 'k',ms=7)\n",
    "        plt.plot([data['TS_'+pv]]*2,[2.1,2.2],'-',color = 'k',ms=7)\n",
    "        pos+=1\n",
    "    plt.text(-1, 2.05,r'$TS$',horizontalalignment='center',verticalalignment='center')\n",
    "    plt.plot([-2,20],[2,2],ls='dotted',color='k')\n",
    "    plt.ylabel('Apparent Vrn Expression')\n",
    "    plt.xlabel('Haun Stage')\n",
    "    plt.ylim(0,2.5)\n",
    "    plt.xlim(-2,18)\n",
    "    plt.text(17,0.05,c,horizontalalignment='right')\n",
    "    if(level>1):\n",
    "        plt.text(-1, 1.05,r'$VI$',horizontalalignment='center',verticalalignment='center')\n",
    "        plt.plot([-2,20],[1,1],ls='dotted',color='k')\n",
    "        #Plot VI_CL treatments at apparent Vrn1 expression of 1.0\n",
    "        plt.plot([data['VI_CL']],[1],'o',mec='b',mfc='b',ms=10)\n",
    "        plt.plot([data['VI_CL']+.1,data['TS_CL']-.1],[1.75,1.75],'>--',color='k')\n",
    "        plt.text((data['VI_CL']+data['TS_CL'])/2,1.75,r'$MinER^{HS}$',horizontalalignment='center',verticalalignment='bottom')\n",
    "        plt.plot([data['VI_CL'],data['VI_CL'],np.nan,data['TS_CL'],data['TS_CL']],[1,1.75,np.nan,1.75,2],ls='dotted',color='k')\n",
    "        plt.text(data['VI_CL'],.67,'CL',horizontalalignment='center',verticalalignment='bottom')\n",
    "        plt.plot([data['VI_CL']],[0.9],'^',color = 'k',ms=7)\n",
    "        plt.plot([data['VI_CL']]*2,[0.9,0.8],'-',color = 'k',ms=7)\n",
    "       #Plot VI_LN treatments at apparent Vrn1 expression of 1.0\n",
    "        plt.plot([data['VI_WL']+.1,data['TS_WL']-.1],[1.55,1.55],'>--',color='k')\n",
    "        plt.text((data['VI_WL']+data['TS_WL'])/2,1.55,r'$MinER^{HS}$',horizontalalignment='center',verticalalignment='bottom')\n",
    "        plt.plot([data['VI_WL'],data['VI_WL'],np.nan,data['TS_WL'],data['TS_WL']],[1,1.55,np.nan,1.55,2],ls='dotted',color='k')\n",
    "        plt.plot([data['VI_WL']],[1],'o',mec='r',mfc='r',ms=10)\n",
    "        plt.text(data['VI_WL'],.67,'WL',horizontalalignment='center',verticalalignment='bottom')\n",
    "        plt.plot([data['VI_WL']],[0.9],'^',color = 'k',ms=7)\n",
    "        plt.plot([data['VI_WL']]*2,[0.9,0.8],'-',color = 'k',ms=7)\n",
    "        if(level>2):    \n",
    "            #Plot VI_CS treatments at apparent Vrn1 expression of 1.0\n",
    "            plt.plot([data['VI_CS']],[1],'o',mec='b',mfc='w',ms=10)\n",
    "            plt.text(data['VI_CS'],0.67,'CS',horizontalalignment='center',verticalalignment='top')\n",
    "            plt.plot([data['VI_CS']+.1,data['TS_CS']-.1],[1.35,1.35],'>-.',color='k')\n",
    "            plt.plot([data['VI_CS'],data['VI_CS'],np.nan,data['TS_CS'],data['TS_CS']],[1,1.35,np.nan,1.35,2],':',color='k')\n",
    "            plt.text((data['VI_CS']+data['TS_CS'])/2,1.35,r'$MaxER^{HS}$',horizontalalignment='center',verticalalignment='bottom')\n",
    "            #Plot VI_WS treatments at apparent Vrn1 expression of 1.0\n",
    "            plt.plot([data['VI_WS']],[1],'o',mec='r',mfc='w',ms=10)\n",
    "            plt.text(data['VI_WS'],0.67,'WS',horizontalalignment='center',verticalalignment='bottom')\n",
    "            plt.plot([data['VI_WS']+.1,data['TS_WS']-.1],[1.15,1.15],'>-.',color='k')\n",
    "            plt.plot([data['VI_WS'],data['VI_WS'],np.nan,data['TS_WS'],data['TS_WS']],[1,1.15,np.nan,1.15,2],':',color='k')\n",
    "            plt.text((data['VI_WS']+data['TS_WS'])/2,1.15,r'$MaxER^{HS}$',horizontalalignment='center',verticalalignment='bottom')\n",
    "            plt.plot([data['VI_WS']],[0.9],'^',color = 'k',ms=7)\n",
    "            plt.plot([data['VI_WS']]*2,[0.9,0.8],'-',color = 'k',ms=7)\n",
    "            if(level>3):\n",
    "                #Plot base Vern1 Rate Vegetative\n",
    "                plt.plot([-data['EmergDurat'],data['VI_WS']],[0,(data['VI_WS']+data['EmergDurat'])*data['rVrnBVeg']],'-',color='g')\n",
    "                MidP = (-data['EmergDurat']+data['VI_WS'])/2\n",
    "                plt.plot([MidP,MidP+1],[0.5,0.5],'--',color='g')\n",
    "                plt.text(MidP+.5,0.53,'rVrnBVeg',color='g')\n",
    "                #Plot max Vrn1 rate vegetative\n",
    "                plt.plot([-data['EmergDurat'],data['VI_CS']],[0,(data['VI_CS']+data['EmergDurat'])*data['MaxDVrnVeg']],'-',color='gold')\n",
    "                MidP = (data['VI_CL']+data['TS_CL'])/2\n",
    "                plt.plot([MidP,MidP+1],[1.5,1.5],'-',color='gold')\n",
    "                plt.text(MidP+.5,1.53,'maxDVrnER',color='gold')\n",
    "                #Plot base Vern1 Rate Early reproductive\n",
    "                plt.plot([data['VI_WS'],data['TS_WS']],[1,1+(data['TS_WS'] - data['VI_WS'])*data['rVrnBER']],'-',color='g')\n",
    "                MidP = (data['VI_WS']+data['TS_WS'])/2\n",
    "                plt.plot([MidP,MidP+1],[1.5,1.5],'--',color='g')\n",
    "                plt.text(MidP+.5,1.53,'rVrnBER',color='g')\n",
    "\n",
    "                #Plot max Vrn1 ER\n",
    "                plt.plot([data['VI_CL'],data['TS_CL']],[1,1+(data['TS_CL'] - data['VI_CL'])*data['MaxDVrnER']],'-',color='gold')\n",
    "                MidP = (data['VI_CL']+data['TS_CL'])/2\n",
    "                plt.plot([MidP,MidP+1],[1.5,1.5],'-',color='gold')\n",
    "                plt.text(MidP+.5,1.53,'MaxDVrnER',color='gold')\n",
    "                if(level>4):\n",
    "                    # Plot end of vernalisation timing\n",
    "                    data['endVern'] = min(data['EndVernTreat'],data['VI_CL'])\n",
    "                    plt.plot([data['endVern']]*2,[-1,2],'--',color='brown')\n",
    "                    plt.text(data['endVern'],-0.1,'endVern',horizontalalignment='center',color='brown')\n",
    "\n",
    "                    #Extrapolate vrn1 back from VI_WL to show when it started effective expression\n",
    "                    plt.plot([data['endVrn2_WL'],data['VI_WL']],[0,data['vrnxVrn3Durat'] * data['dvrnxVrn3']],'--',color='grey')\n",
    "                    MidP = data['endVrn2_WL']+data['vrnxVrn3Durat']*0.1\n",
    "                    plt.plot([MidP,MidP+1],[0.1]*2,'--',color='grey')\n",
    "                    plt.text(MidP+.2,0.15,'dvrnxVrn3',color='grey')\n",
    "\n",
    "                    #Extrapolate BaseVrn1 back to MinVI to show how much Vrn1 was required to overcome Vrn2\n",
    "                    plt.plot([0,data['endVrn2_WL']],[data['rVrn2'],0],'--',color='Orange')\n",
    "                    plt.text(0,min(2.15,data['rVrn2']),'Vrn2',color='orange')\n",
    "\n",
    "                    #The amount of Vrn1 required to reach VI is light is 1 + rVrn2\n",
    "                    plt.plot([0,data['VI_CL']],[1+data['rVrn2']]*2,'--',color='k')\n",
    "                    plt.text(data['VI_CL']-.2,1+data['rVrn2']+.03,'Vrn Target',horizontalalignment='right',verticalalignment='bottom')\n",
    "                    #the amount required from cold up regulation is the above less base Vrn1 until the end of Vrn2 less base * Vrn3 between end vrn2 and VI\n",
    "                    plt.plot([0,data['VI_CL']],[1+data['rVrn2']-data['VrnBVeg_CL']]*2,'-',color='g')\n",
    "                    plt.text(data['VI_CL']-.2,data['Vrn1_CL']+.03,'less baseVrn',horizontalalignment='right',verticalalignment='bottom')\n",
    "\n",
    "                    #RelMethTime = camp.CampConstants['MethalationThreshold']/(data['Vrn1_CL']+camp.CampConstants['MethalationThreshold'])\n",
    "                    StartMethHS = -data['EmergDurat'] + data['Vrn1HS_CL'] * 0.5 \n",
    "                    plt.plot([StartMethHS,data['endVern']],[0,data['Vrn1_CL']],'-',color='cyan')\n",
    "                    plt.text(-data['EmergDurat'],data['Vrn1_CL'],'Vrn1_CL',horizontalalignment='right',rotation=90,verticalalignment='top',color='c')\n",
    "    \n",
    "def boundPlots(data, ax, c):\n",
    "    lowestTS = 16\n",
    "    highestTS = 0\n",
    "    pos = 0\n",
    "    \n",
    "    #Plot TS for all treats at apparent Vrn1 expression of 2.0\n",
    "    for pv in ['WS','CL','CS','WL']:\n",
    "        #plt.plot(CampInputs.loc[c,'FLN_'+pv],2.5,'*',mfc = fills[pos],mec=colors[pos],ms=10)\n",
    "        plt.plot([data['TS_'+pv]],[2],'o',mfc = fills[pos],mec=colors[pos],ms=10)\n",
    "        plt.text(data['TS_'+pv],2.25,pv,horizontalalignment='center')\n",
    "        lowestTS = min(lowestTS,data['TS_'+pv])\n",
    "        highestTS = max(highestTS,data['TS_'+pv])\n",
    "        plt.plot([data['TS_'+pv]],[2.1],'v',color = 'k',ms=7)\n",
    "        plt.plot([data['TS_'+pv]]*2,[2.1,2.2],'-',color = 'k',ms=7)\n",
    "        pos+=1\n",
    "    plt.text(-1, 2.05,r'$TS$',horizontalalignment='center',verticalalignment='center')\n",
    "    plt.plot([-2,20],[2,2],ls='dotted',color='k')\n",
    "    plt.ylabel('Apparent Vrn Expression')\n",
    "    plt.xlabel('Haun Stage')\n",
    "    plt.ylim(0,2.5)\n",
    "    plt.xlim(-2,18)\n",
    "    plt.text(17,0.05,c,horizontalalignment='right')\n",
    "    \n",
    "    plt.text(-1, 1.05,r'$VI$',horizontalalignment='center',verticalalignment='center')\n",
    "    plt.plot([-2,20],[1,1],ls='dotted',color='k')\n",
    "    #Plot VI_CL treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_CL']],[1],'o',mec='b',mfc='b',ms=10)\n",
    "    plt.text(data['VI_CL'],1.3,'CL',horizontalalignment='center',verticalalignment='bottom')\n",
    "    plt.plot([data['VI_CL']],[1.1],'v',color = 'k',ms=7)\n",
    "    plt.plot([data['VI_CL']]*2,[1.1,1.2],'-',color = 'k',ms=7)\n",
    "   #Plot VI_WL treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_WL']],[1],'o',mec='r',mfc='r',ms=10)\n",
    "    plt.text(data['VI_WL'],.67,'WL',horizontalalignment='center',verticalalignment='bottom')\n",
    "    plt.plot([data['VI_WL']],[0.9],'^',color = 'k',ms=7)\n",
    "    plt.plot([data['VI_WL']]*2,[0.9,0.8],'-',color = 'k',ms=7)\n",
    "    \n",
    "    #Plot VI_CS treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_CS']],[1],'o',mec='b',mfc='w',ms=10)\n",
    "    plt.text(data['VI_CS'],1.3,'CS',horizontalalignment='center',verticalalignment='top')\n",
    "    \n",
    "    #Plot VI_WS treatments at apparent Vrn1 expression of 1.0\n",
    "    plt.plot([data['VI_WS']],[1],'o',mec='r',mfc='w',ms=10)\n",
    "    plt.text(data['VI_WS'],0.67,'WS',horizontalalignment='center',verticalalignment='bottom')\n",
    "    # plt.plot([data['VI_WS']+.1,data['TS_WS']-.1],[1.15,1.15],'>-.',color='k')\n",
    "    # plt.plot([data['VI_WS'],data['VI_WS'],np.nan,data['TS_WS'],data['TS_WS']],[1,1.15,np.nan,1.15,2],':',color='k')\n",
    "    # plt.text((data['VI_WS']+data['TS_WS'])/2,1.15,r'$MaxER^{HS}$',horizontalalignment='center',verticalalignment='bottom')\n",
    "    plt.plot([data['VI_WS']],[0.9],'^',color = 'k',ms=7)\n",
    "    plt.plot([data['VI_WS']]*2,[0.9,0.8],'-',color = 'k',ms=7)\n",
    "            \n",
    "    basepos = 0.25\n",
    "    #Plot base Vern1 Rate Vegetative\n",
    "    plt.plot([-data['EmergDurat'],data['VI_WS']],[0,(data['VI_WS']+data['EmergDurat'])*data['rVrnBVeg']],'-',color='g')\n",
    "    MidP = ((data['EmergDurat']+data['VI_WS'])* basepos)-data['EmergDurat']\n",
    "    ang = np.arctan(1/(data['EmergDurat']+data['VI_WS']))*(180/np.pi)\n",
    "    arc = patches.Arc([MidP,basepos],1.0,.125,angle=0.0,theta1=0.0,theta2=ang,linewidth=2, fill=False, color='g')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+1],[basepos]*2,'-',color='g')\n",
    "    plt.text(MidP+.7,basepos+.03,r'$rVrnB_{Veg}$',color='g')\n",
    "    \n",
    "    maxpos = 0.75\n",
    "    #Plot max Vrn1 rate vegetative\n",
    "    plt.plot([-data['EmergDurat'],data['VI_CS']],[0,(data['EmergDurat']+data['VI_CS'])*data['rApDevMVeg']],'-',color='gold')\n",
    "    MidP = ((data['EmergDurat']+data['VI_CS']) * maxpos)-data['EmergDurat']\n",
    "    ang = np.arctan(1/(data['EmergDurat']+data['VI_CS']))*(180/np.pi)\n",
    "    arc = patches.Arc([MidP,maxpos],1.0,.125,angle=0.0,theta1=0.0,theta2=ang,linewidth=2, fill=False, color='gold')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+1],[maxpos]*2,'-',color='gold')\n",
    "    plt.text(MidP+.7,maxpos+0.03,r'$rApDevM_{Veg}$',color='gold', horizontalalignment = 'left')\n",
    "    \n",
    "    #Plot base Vern1 Rate Early reproductive\n",
    "    plt.plot([data['VI_WS'],data['TS_WS']],[1,1+(data['TS_WS'] - data['VI_WS'])*data['rVrnBER']],'-',color='g')\n",
    "    MidP =  data['VI_WS'] + (data['VI->TS_CS'] * basepos)\n",
    "    ang = np.arctan(1/data['VI->TS_CS'])*(180/np.pi)\n",
    "    arc = patches.Arc([MidP,1+basepos],1.0,.125,angle=0.0,theta1=0.0,theta2=ang,linewidth=2, fill=False, color='g')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+1],[1+basepos]*2,'-',color='g')\n",
    "    plt.text(MidP+.7,1+basepos+.03,r'$rVrnB_{ER}$',color='g')\n",
    "\n",
    "    # #Plot max Vrn1 ER\n",
    "    plt.plot([data['VI_CL'],data['TS_CL']],[1,1+(data['TS_CL'] - data['VI_CL'])*data['rApDevMER']],'-',color='gold')\n",
    "    MidP = data['VI_CL']+(data['MinVI->TS']* maxpos)\n",
    "    ang = np.arctan(1/(data['EmergDurat']+data['VI_CS']))*(180/np.pi)\n",
    "    arc = patches.Arc([MidP,1+maxpos],1.0,.125,angle=0.0,theta1=0.0,theta2=ang,linewidth=2, fill=False, color='gold')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+1],[1+maxpos]*2,'-',color='gold')\n",
    "    plt.text(MidP+.7,1+maxpos+0.03,r'$rApDevM_{ER}$',color='gold',)\n",
    "    \n",
    "    plt.plot(-data['EmergDurat'],0.03,'v',color = 'k',ms=7)\n",
    "    plt.plot([-data['EmergDurat']]*2,[0,.2],'-',color = 'k',ms=7)\n",
    "    plt.text(-data['EmergDurat'],.25,'Imbibation', rotation = 90, verticalalignment='bottom',horizontalalignment = 'center')\n",
    "    \n",
    "    plt.plot(0,0.03,'v',mec = 'k',mfc='white',ms=7)\n",
    "    plt.plot([0]*2,[.08,.4],'--',color = 'k',ms=7)\n",
    "    plt.text(0,.45,'Emergence', rotation = 90, verticalalignment='bottom',horizontalalignment = 'center')\n",
    "    \n",
    "def Vrn1Plots(data,ax, c):\n",
    "    lowestTS = 16\n",
    "    highestTS = 0\n",
    "    pos = 0\n",
    "\n",
    "    plt.ylabel('Apparent Vrn Expression')\n",
    "    plt.xlabel('Haun Stage')\n",
    "    plt.ylim(0,1.8)\n",
    "    plt.xlim(-1.2,5)\n",
    "    plt.text(5,0.05,c,horizontalalignment='right')\n",
    "    \n",
    "    plt.text(4.8, 1.05,r'$VI$',horizontalalignment='center',verticalalignment='center')\n",
    "    plt.plot([-2,20],[1,1],ls='dotted',color='k')\n",
    "\n",
    "    plt.plot([data['VI_CS']],[1],'o',mec='b',mfc='b',ms=10)\n",
    "    plt.text(data['VI_CS'],.67,'CS',horizontalalignment='center',verticalalignment='bottom')\n",
    "    plt.plot([data['VI_CS']],[0.9],'^',color = 'k',ms=7)\n",
    "    plt.plot([data['VI_CS']]*2,[0.9,0.8],'-',color = 'k',ms=7)\n",
    " \n",
    "    #Extrapolate BaseVrn1 back to MinVI to show how much Vrn1 was required to overcome Vrn2\n",
    "    plt.text(data['VI_CS']+.92,(2+data['rVrn2'])/2,r'$rVrn2$',color='crimson')\n",
    "    plt.plot([data['VI_CS']+.9]*2,[1.02,0.98+data['rVrn2']], '^--', color='crimson')\n",
    "    \n",
    "    basepos = 0.12\n",
    "    #Plot base Vern1 Rate Vegetative\n",
    "    plt.plot([-data['EmergDurat'],data['VI_CS']],[0,(data['VI_CS']+data['EmergDurat'])*data['rVrnBVeg']],'-',color='g')\n",
    "    MidP = ((data['EmergDurat']+data['VI_WS'])* basepos)-data['EmergDurat']\n",
    "    ang = np.arctan(1/(data['EmergDurat']+data['VI_WS']))*(180/np.pi)\n",
    "    arc = patches.Arc([MidP,basepos],1.0,.125,angle=0.0,theta1=0.0,theta2=ang,linewidth=2, fill=False, color='g')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+1],[basepos]*2,'-',color='g')\n",
    "    plt.text(MidP+.7,basepos+.03,r'$rVrnB_{Veg}$',color='g')\n",
    "    \n",
    "    plt.plot([data['VI_CS'],data['VI_CS']],[0.02,(data['VI_CS']+data['EmergDurat'])*data['rVrnBVeg']-.02],'^-.',color='g')\n",
    "    plt.text(data['VI_CS']+.02,0.2,r'$\\Sigma^0_{VI}VrnB_{Veg}$',horizontalalignment='left',verticalalignment='bottom',color='g')\n",
    "    \n",
    "    plt.plot([1.5,1.5],[1.02+data['rVrn2']-data['VrnBVeg_CL'],data['rVrn2']+.98],'^-.',color='g')\n",
    "    plt.text(1.52,1.2,r'$\\Sigma^0_{VI}VrnB_{Veg}$',horizontalalignment='left',verticalalignment='bottom',color='g')\n",
    "    \n",
    "    plt.plot([-2,data['VI_CS']+.9],[1+data['rVrn2']]*2,'--',color='k')\n",
    "    plt.text(-1,1.02+data['rVrn2'],r'$Vrn^PVI_{CL}$')\n",
    "    \n",
    "    plt.plot([-2,1.5],[1+data['rVrn2']-data['VrnBVeg_CL']]*2,'-.',color='k')\n",
    "    plt.text(-1,1.02+data['rVrn2']-data['VrnBVeg_CL'],'$Vrn1_{CL}$')\n",
    "    \n",
    "    plt.plot(0,0.03,'v',mec = 'b',mfc='b',ms=7)\n",
    "    plt.plot([0,0,np.nan,0,0,np.nan,0,0],[.03,.1,np.nan,.45,.55,np.nan,.7,1.0+data['rVrn2']-data['VrnBVeg_CL']],'--',color = 'b',ms=7)\n",
    "    plt.text(0,.15,'End Vern', rotation = 90, verticalalignment='bottom',horizontalalignment = 'center',color='b')\n",
    "    \n",
    "    MethHS = 1/(1+1.+data['rVrn2']-data['VrnBVeg_CL'])\n",
    "    Vrn1Needed = 1.0+data['rVrn2']-data['VrnBVeg_CL']\n",
    "    basepos = 0.6\n",
    "    plt.plot([-data['EmergDurat']*MethHS,0],[0,Vrn1Needed],'-',color='b')\n",
    "    MidP = (-data['EmergDurat']*MethHS)/2 \n",
    "    ang = np.arctan((Vrn1Needed)/(data['EmergDurat']*MethHS))*(180/np.pi)\n",
    "    arc = patches.Arc([MidP,basepos],0.5,.125,angle=0.0,theta1=0.0,theta2=ang,linewidth=2, fill=False, color='b')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+.5],[basepos]*2,'-',color='b')\n",
    "    plt.text(MidP+0.35,basepos+.05,r'$\\Delta Vrn1_{CL}$',color='b')\n",
    "    ang = np.arctan(((Vrn1Needed))/(data['EmergDurat']*MethHS))*(180/np.pi)*1.23\n",
    "    plt.text(-data['EmergDurat']*MethHS+.04,.15,'Methalated Vrn1', rotation = ang, verticalalignment='bottom',horizontalalignment = 'center',color='b')\n",
    "        \n",
    "    plt.plot([-data['EmergDurat'],-data['EmergDurat']*MethHS],[0,1.0],'-',color='deepskyblue')\n",
    "    plt.text(-data['EmergDurat']+.1,.15,'UnMethalated Vrn1', rotation = ang, verticalalignment='bottom',horizontalalignment = 'center',color='deepskyblue')\n",
    "    \n",
    "    plt.text(5,0.92,'MT',color='k',horizontalalignment = 'right')\n",
    "    \n",
    "def Vrn2Plots(data, ax, c):\n",
    "    lowestTS = 16\n",
    "    highestTS = 0\n",
    "    pos = 0\n",
    "\n",
    "    plt.ylabel('Apparent Vrn Expression')\n",
    "    plt.xlabel('Haun Stage')\n",
    "    plt.ylim(0,2.5)\n",
    "    plt.xlim(-2,18)\n",
    "    plt.text(17,0.05,c,horizontalalignment='right')\n",
    "    \n",
    "    plt.text(-1, 1.05,r'$VI$',horizontalalignment='center',verticalalignment='center')\n",
    "    plt.plot([-2,20],[1,1],ls='dotted',color='k')\n",
    "\n",
    "    plt.plot([data['VI_WL']],[1],'o',mec='r',mfc='r',ms=10)\n",
    "    plt.text(data['VI_WL'],.67,'WL',horizontalalignment='center',verticalalignment='bottom')\n",
    "    plt.plot([data['VI_WL']],[0.9],'^',color = 'k',ms=7)\n",
    "    plt.plot([data['VI_WL']]*2,[0.9,0.8],'-',color = 'k',ms=7)\n",
    " \n",
    "    #Extrapolate BaseVrn1 back to MinVI to show how much Vrn1 was required to overcome Vrn2\n",
    "    plt.plot([0,data['DRVrn2HSWL']],[data['rVrn2'],0],'--',color='g')\n",
    "    plt.text(-.5,min(2.15,data['rVrn2'])+.02,R'$rVrn2$',color='crimson')\n",
    "    plt.plot([-.5,2.5],[data['rVrn2']]*2, '-', color='crimson')\n",
    "    \n",
    "\n",
    "    # #Plot max Vrn1 ER\n",
    "    maxpos = 0.5\n",
    "    MidP = (data['DRVrn2HSWL']* maxpos)\n",
    "    if data['DRVrn2HSWL']!=0:\n",
    "        ang = np.arctan(data['rVrn2']/(data['DRVrn2HSWL']))*(180/np.pi)\n",
    "    else:\n",
    "        ang = 0\n",
    "    arc = patches.Arc([MidP,data['rVrn2']*(1-maxpos)],1.0,.125,angle=0.0,theta1=360-ang,theta2=0,linewidth=2, fill=False, color='g')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+1],[data['rVrn2']*(1-maxpos)]*2,'-',color='g')\n",
    "    plt.text(MidP+.6,data['rVrn2']*(1-maxpos)-0.08,r'$-(rVrnB_{Veg})$',color='g',)\n",
    "\n",
    "    plt.plot([data['DRVrn2HSWL'],data['VI_WL']],[0,data['URVrn3HSWL']*(data['rVrnBVeg'] * data['rVrn3Veg'])],'--',color='grey')\n",
    "    maxpos = 0.5\n",
    "    MidP = data['DRVrn2HSWL']+(data['VI_WL']-data['DRVrn2HSWL'])* maxpos\n",
    "    ang = np.arctan(1/(data['VI_WL']-data['DRVrn2HSWL']))*(180/np.pi)\n",
    "    arc = patches.Arc([MidP,1*maxpos],1.0,.125,angle=0.0,theta1=0,theta2=ang,linewidth=2, fill=False, color='grey')\n",
    "    ax.add_patch(arc)\n",
    "    plt.plot([MidP,MidP+1],[maxpos]*2,'-',color='grey')\n",
    "    plt.text(MidP+.6,maxpos+0.03,r'$rVrnB_{Veg} * rVrn3_{Veg}$',color='grey',)\n",
    "    plt.plot([data['VI_WL']]*2, [1.0,1.15],':',color='k')\n",
    "    plt.plot([data['DRVrn2HSWL']+.1,data['DRVrn2HSWL']+data['URVrn3HSWL']],[1.15,1.15],'>-.',color='k')\n",
    "    plt.text(data['DRVrn2HSWL']+data['URVrn3HSWL']/2,1.15,r'$URVrn3^{HS}_{WL}$',horizontalalignment='center',verticalalignment='bottom')\n",
    "    \n",
    "    plt.plot(0,0.03,'v',mec = 'k',mfc='white',ms=7)\n",
    "    plt.plot([0,0,np.nan,0,0,np.nan,0,0],[.06,.1,np.nan,0.45,data['rVrn2'],np.nan,data['rVrn2']+.1,1.],'-',color = 'k',ms=7)\n",
    "    plt.text(0,.15,'Emerg', rotation = 90, verticalalignment='bottom',horizontalalignment = 'center')\n",
    "    \n",
    "    plt.plot([0+.1,data['DRVrn2HSWL']],[1.15,1.15],'>-.',color='k')\n",
    "    plt.plot([0,0,np.nan,data['DRVrn2HSWL'],data['DRVrn2HSWL']],[1,1.15,np.nan,1.15,0],':',color='k')\n",
    "    plt.text(data['DRVrn2HSWL']/2,1.15,r'$DRVrn2^{HS}_{WL}$',horizontalalignment='center',verticalalignment='bottom')      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
